### DISCUSSION TITLE
MUI Autocomplete Problems.

### DISCUSSION THREAD

**Discussion by RealMarvn** [react-hook-form/react-hook-form#8157] [Q&A]
Heyho,
I want to have a Autocomplete component by MUI wrapped with the Controller as a component for my future FORMs.
I ran into some issues regarding arrays (if I set the autocomplete to multiple) and some issues itself on selecting some stuff.
I am gonna send a codepen in here if that's ok.
Thanks for the help in advance.
I am using TypeScript and React with React Hook Form.
Best Regards. Marvin

**Comment by RealMarvn** [comment]
https://codesandbox.io/s/optimistic-elbakyan-08puvl?file=/src/FormInputSelect.tsx

**Reply by RealMarvn** [reply]
I made this public and you can edit it.

**Reply by Moshyfawn** [reply]
Looks like u didn't save your codesanbox

<img width="1050" alt="image" src="https://user-images.githubusercontent.com/16290753/162613319-f496ef95-d73a-4ad2-bace-bcedbf1120ad.png">

**Reply by RealMarvn** [reply]
Can you reload? It should now be visible

**Reply by RealMarvn** [reply]
I am pretty new to all this stuff but I stuck like never before. Please help me

**Comment by RealMarvn** [comment]
```import React from 'react'
import {Controller, useFormContext} from "react-hook-form";
import {Autocomplete, TextField} from "@mui/material";

interface IOption {
    name: string
    value: string
}

export const FormInputSelectV2 = (props: { name: string, label: string, optionValues: Array<IOption>, defaultValue: IOption }) => {

    const {control} = useFormContext()

    return (
        <Controller
            name={props.name}
            control={control}
            rules={{
                required: "Please select a value"
            }}
            defaultValue={props.defaultValue}
            render={({field: {onChange, value}, fieldState: {error}, formState}) => (
                <Autocomplete
                    onChange={onChange}
                    value={value}
                    // multiple
                    options={props.optionValues}
                    getOptionLabel={(option) => option.name}
                    filterSelectedOptions
                    renderInput={(params) => <TextField {...params} label={props.label}/>}
                />
            )}
        />
    )
}```

**Reply by RealMarvn** [reply]
This is how the render method actually looks like. Didnt work on CodeSandBox but here you have it. I want to make it return a single IOption interface value if multiple is off and if multiple is on , it should be able to return an array as IOption. Multiple should be a prop at the top sometimes so i can activate it if I need that.

**Reply by bluebill1049** [reply]
is this example helps? https://codesandbox.io/s/react-hook-form-v7-controller-5h1q5?file=/src/MuiAutoComplete.js

**Reply by RealMarvn** [reply]
Nope. I already looked at that example 200 times. If you change the autocomplete to multiple , it destroys everything

**Comment by Moshyfawn** [comment]
[Here you go](https://codesandbox.io/s/hungry-johnson-2mu824?file=%2Fsrc%2FDemo.tsx). The types are a bit messy and there's a ton of room for improvement, but the input's working as intended _(I believe)_

_P.S. I was already working on [a codessanbox](https://codesandbox.io/s/hungry-johnson-2mu824?file=%2Fsrc%2FDemo.tsx), so it's not [your specific example](https://codesandbox.io/s/optimistic-elbakyan-08puvl?file=/src/FormInputSelect.tsx), but the same functionality_

**Reply by RealMarvn** [reply]
Does it also work if I add multiple? Most of the examples are breaking if I change that var

**Reply by RealMarvn** [reply]
Oh i just saw the example code. This is so much code. Damn hahaha. Thanks for your help.

**Reply by RealMarvn** [reply]
Uhm i just looked at your code and I am a bit confused. As I am said I am a bit new to all of this.  What exactly did you do different to make it work?

**Reply by Moshyfawn** [reply]
> Uhm i just looked at your code and I am a bit confused. As I am said I am a bit new to all of this. What exactly did you do different to make it work?

The main issue you where having was with the args MUI Autocomplete returns `onChange`. Basically, you were sending the `event` object into your field's `onChange` handler, which works for simple values while the output is not an array, as it is with `multiple` prop set. MUI Autocomplete returns `event` first and then the actual `value`, which you want to pass into your field's `onChange` handler.

```diff
- onChange={onChange}
+ onChange={(evt, value) => onChange(value)}
```

**Reply by RealMarvn** [reply]
ok let me change that on my code. Give me a second

**Reply by Moshyfawn** [reply]
> ok let me change that on my code. Give me a second

For sure, no rush

**Reply by RealMarvn** [reply]
Oh crap. Now nothing renders and there are three errors

**Reply by RealMarvn** [reply]
![image](https://user-images.githubusercontent.com/67916561/162615544-6be2a659-f91b-4517-8764-fba9dbee78f1.png)
![image](https://user-images.githubusercontent.com/67916561/162615551-ba126e3d-c468-4939-9a34-4853ca7599a3.png)

**Reply by Moshyfawn** [reply]
Wait, so what version of RHF are you working with?

[You codesandbox is is using `v6.12.2`](https://codesandbox.io/s/optimistic-elbakyan-08puvl?file=%2Fpackage.json%3A377-410)

```json
"dependencies": {
  "react-hook-form": "6.12.2"
}
```

**Reply by RealMarvn** [reply]
> 

I am so sry. I am using 7.29.0 ..... I forgot to change that in the CodePen

**Reply by Moshyfawn** [reply]
No worries!
>  I forgot to change that in the CodePen

This might be the cause of those 3 errors

**Reply by RealMarvn** [reply]
oh...Should I rollback to that older version or is there a new fix?

**Reply by Moshyfawn** [reply]
You're using a V7 API with V6 RHF version. And the `Controller` API has had some breaking changes.

If you're using V7 in your project, just update your codesanbox to V7

**Reply by RealMarvn** [reply]
I updated to 7 in my Codepen....There are now much more errors.

**Reply by Moshyfawn** [reply]
Are you using [this codesanbox](https://codesandbox.io/s/optimistic-elbakyan-08puvl?file=%2Fsrc%2FFormInputSelect.tsx)? I'm confused 'cuz [you've posted your `FormInputSelectV2` code](https://github.com/react-hook-form/react-hook-form/discussions/8157#discussioncomment-2538873) and it's different from the one I see in [this codesanbox](https://codesandbox.io/s/optimistic-elbakyan-08puvl?file=%2Fsrc%2FFormInputSelect.tsx)

**Reply by RealMarvn** [reply]
Yeah the only thing that changed to my custom code and this codesandbox is that i define those redner vars.
`` render={({field: {onChange, value}, fieldState: {error}, formState}) => (`` 
But i can also switch to that params thing on my main code. Shouldnt be a problem for me. I ve just seen both versions and I dont know which would be better. But except for that change they should be equal.

**Reply by Moshyfawn** [reply]
Well, spreading `{...parameters}` results in passing the `field`, `fieldState` and `formState` props into the MUI Autocomplete instead of the actual handlers and the value from the `field` prop, doesn't it?

**Reply by RealMarvn** [reply]
Ah so i should just change those fields into props because its the same?

**Reply by Moshyfawn** [reply]
Alright, I got it.

You are setting your default value to an object `channelOptions[0]` and MUI Autocomplete with `multiple` set expects an array.

```diff
- defaultValue={channelOptions[0]}
+ defaultValue={[channelOptions[0]]}
```

**Reply by Moshyfawn** [reply]
Here's an example from [your codesanbox](https://codesandbox.io/s/trusting-hofstadter-dwfsyj?file=%2Fsrc%2FFormInputSelect.tsx)

**Reply by RealMarvn** [reply]
> Here's an example from [your codesanbox](https://codesandbox.io/s/trusting-hofstadter-dwfsyj?file=%2Fsrc%2FFormInputSelect.tsx)

Did you make those changes already there?

**Reply by RealMarvn** [reply]
Ahhhh ok. I just looked at that code and its working now! Great! Uhm but now there is another problem....If I change that autocomplete to only one value (I removed multiple as a prop) then its throwing errors...

**Reply by Moshyfawn** [reply]
> Ah so i should just change those fields into props because its the same?

You want to pass the value and the handlers into their appropriate props. You can check out[ the RHF `Controller` docs](https://react-hook-form.com/api/usecontroller/controller) to learn what args does the render callback return.

So, basically, something like this:
```diff
- render={(parameters) =>
+ render={({ field: { onChange, ref, name, ...fieldProps } }) =>
     <Autocomplete
-     {...parameters}
+     {...fieldProps}
+     onChange={(evt, value) => onChange(value)}
       renderInput={(params) => (
         <TextField
           {...params}
           label={props.label}
+          name={name}
+          inputRef={ref}
         />
       )}
     />
 }
```

You can destruct the `ref`, the `name` and the `onChange` handler and pass them correctly as the MUI Autocomplete API is different from the RHF field params shape and spread the rest of the field props _(`onBlur`, `value`)_ as they are the same.

Same goes to the `fieldState` props like `invalid`, `error` and whatever else form info you wanna use with your input: just destruct the `fieldState` prop from the `Controller`'s `render` callback params and pass the correct states into the appropriate MUI Autocomplete props.

For example:

```js
render={({ fieldState: { invalid, error } }) => (
  <Autocomplete
    error={invalid}
    helperText={error?.message}
  />
)}
```

**Reply by Moshyfawn** [reply]
> Uhm but now there is another problem....If I change that autocomplete to only one value (I removed multiple as a prop) then its throwing errors...

Well, you gotta pass the appropriate `defaultValue` depending on the type of your input: an `object` _(or a `string`)_ to a single value Autocomplete _(without the `multiple` prop set)_ and an `array` of `object`s, if `multiple` prop is set.

It's the MUI Autocomplete API and RHF has nothing to do with it.

**Reply by bluebill1049** [reply]
thanks, @Moshyfawn's detailed help. Most of the time, I always make the controlled component first (so understand the component API) then wrapped/hooked with Controller/useController.

**Reply by RealMarvn** [reply]
Ahhhh now I understand! great!

**Reply by RealMarvn** [reply]
Cool. I am gonna play around with that a bit and ask again if I dont understand something! Great!
I also got a question to react router but that does not belong here hahahaha.. Or am I allowed to ask right here if there is already someone who helps me?
Thanks for your help

**Reply by RealMarvn** [reply]
Hey got another question...How can I set no value in defaultValue?

**Reply by RealMarvn** [reply]
If i remove the defaultValue prop completly , it throws an error cause of the undefined state to a defined state and so on. I am kinda confused rn

**Reply by Moshyfawn** [reply]
> Cool. I am gonna play around with that a bit and ask again if I dont understand something! Great!

Sounds good, let me know if you get stuck

> I also got a question to react router but that does not belong here hahahaha.. Or am I allowed to ask right here if there is already someone who helps me? 

It's better to ask your react-router related question over in [their repository discussions section](https://github.com/remix-run/react-router/discussions). We here only maintain the RHF package.

You can leave a link to your discussion here once you've posted it; I might take a look and try helping you, tho, I don't help maintain the react-router package

**Reply by RealMarvn** [reply]
> 

I asked right there on -> https://github.com/remix-run/react-router/discussions/8787
Thanks btw

**Reply by Moshyfawn** [reply]
> If i remove the defaultValue prop completly , it throws an error cause of the undefined state to a defined state and so on. I am kinda confused rn

Depending on the input _(single / multiple)_, you want to provide either an empty `array` _(if multiple)_ or `null`, if it's a single value input
<img width="274" alt="image" src="https://user-images.githubusercontent.com/16290753/162618210-19e8e7de-3e81-480e-b994-df6f3c22c30a.png">

**Reply by RealMarvn** [reply]
ok thx

**Comment by RealMarvn** [comment]
```import React from 'react'
import {Controller, useFormContext} from "react-hook-form";
import {Autocomplete, TextField} from "@mui/material";

interface IOption {
    name: string
    value: string
}

export const FormInputSelectV2 = (props: { name: string, label: string, optionValues: Array<IOption>, defaultValue: IOption }) => {

    const {control} = useFormContext()

    return (
        <Controller
            name={props.name}
            control={control}
            rules={{
                required: "Please select a value"
            }}
            defaultValue={props.defaultValue}
            render={({field: {onChange, value}, fieldState: {error}, formState}) => (
                <Autocomplete
                    onChange={(evt, value) => onChange(value)}
                    value={value}
                    multiple
                    options={props.optionValues}
                    getOptionLabel={(option) => option.name}
                    filterSelectedOptions
                    renderInput={(params) => <TextField {...params} label={props.label}/>}
                />
            )}
        />
    )
}
``` This is my code currently

**Comment by RealMarvn** [comment]
Hey guys got another question haha

**Reply by RealMarvn** [reply]
```
        <div className="welcomer__message">
            <FormProvider {...methods}>
                <FormHeader header="Message" switchName="statusBoolean" switchDefaultValue={false}/>
                <hr/>

                <div className="control-body">
                    <section>
                        <p>Channel where your message will be sent</p>
                        <FormInputSelect name="channelValue" label="Select your channel" optionValues={channelOptions}
                                         defaultValue={channelOptions[0]} multiple={false}/>
                    </section>
                    <hr/>
                    <section>
                        <p>Your text message</p>
                        <FormInputText name="messageValue" label="Message" defaultValue="" rows={3}/>
                    </section>
                    <hr/>
                    <section>
                        <p>Theme which will be used</p>
                        <FormInputSelect name="themeValue" label="Select your Theme" optionValues={themeOptions}
                                         defaultValue={themeOptions[0]} multiple={false}/>
                    </section>

                    <FormFooter disabled={!methods.formState.isDirty || methods.formState.isSubmitting}
                                handleSubmit={handleSubmit(onSubmit)}
                                reset={true}
                                handleReset={() => reset()}
                    />
                </div>
            </FormProvider>
        </div>```

This is my Form....I want to build a preview image where the text input from the  text form component will be shown in an image. I already know how to generate such an image but i dont know how i get the value of my messageValue component
